add generic syrk/herk

araujoms opened the pull request:
I've added an implementation of syrk/herk for generic types, in order to avoid falling back to `_generic_matmatmul!`, as it's rather slow. I didn't do anything fancy, no multithreading or anything, but this gives a 1.5x to 2x speedup for `Int` and `BigFloat`, for example.

The generic version of syrk only works for real and complex numbers, but funnily enough herk works for anything that respects `conj(a*b) == conj(b)*conj(a)`, which as far as I can tell is any subtype of `Number`, including quaternions and octonions.

I've ran the tests locally by reverting to the commit before the lazy JLLs one, and they pass.

araujoms committed:
add generic syrk/herk

araujoms committed:
add vector

araujoms committed:
typo

araujoms committed:
typo

dkarrasch commented on src/matmul.jl:
```diff
@@ -570,6 +570,89 @@ Base.@constprop :aggressive function _symm_hemm_generic!(C, tA, tB, A, B, alpha,
+    end
+    @inbounds if !conjugate
+        if aat
+            for k ∈ 1:n, j ∈ 1:m
```
Unless the `StridedArray` type guarantees that you may need to check `requires_one_based_indexing(A, C)` somewhere upfront.

araujoms replied:
you're right, thanks for catching it

dkarrasch commented on src/matmul.jl:
```diff
@@ -570,6 +570,89 @@ Base.@constprop :aggressive function _symm_hemm_generic!(C, tA, tB, A, B, alpha,
+        fill!(C, T(0))
+    elseif !isone(β)
+        C .*= β
+    end
```
Isn't this 
```suggestion
    _rmul_or_fill!(C, β)
```
?

araujoms replied:
Well, almost, `rmul!` doesn't skip the multiplication when β is one. I guess it makes more sense to add this test there than to have an ad hoc implementation here, though.

dkarrasch replied:
That sounds very reasonable.

dkarrasch commented on src/matmul.jl:
```diff
@@ -570,6 +570,89 @@ Base.@constprop :aggressive function _symm_hemm_generic!(C, tA, tB, A, B, alpha,
     _generic_matmatmul!(C, wrap(A, tA), wrap(B, tB), alpha, beta)
 end
 
+Base.@constprop :aggressive function generic_matmatmul_wrapper!(C::StridedMatrix{T}, tA, tB, A::StridedVecOrMat{T}, B::StridedVecOrMat{T},
```
Maybe a stupid thought, but do you really need all three eltypes to be equal? Could they be mixed (in any sense, including unitful/unitless etc.)? And, actually, even non-`Number`? That would require replacing `conj` by `adjoint` and `abs2` by the product (something like `A[j,k]'A[j,k]` or so?) below.

araujoms replied:
Well if the type of `A` and `B` are not the same `A === B` will fail and `generic_syrk!` is never going to get called. The type of `C` can be different, if it can handle the type of `A`, but I don't see the benefit. This function is rather low-level, I'd rather enforce the same type, as it will have better performance.

We can have non-`Number` indeed, as was in fact what I first wrote (also requires putting a `transpose` in the syrk version and changing `fill!(C, T(0))` to `fill!(C, zero(first(A)))`. But then I thought that we already have `_generic_matmatmul!` to handle the fancy types, it's better to focus here on performance for plain numerical matrices. `abs2(z)` is more efficient than `z*conj(z)`, and moreover guarantees that the diagonal elements are exactly real (which is otherwise not necessarily true, see JuliaMath/DoubleFloats.jl#211).

dkarrasch replied:
Sure, if needed this can always be relaxed.

dkarrasch reviewed (commented):
Cool, better performance is always welcome. I've got some comments to explore the opportunity to make it even more generic.

araujoms committed:
_rmul_or_full!

Co-authored-by: Daniel Karrasch <daniel.karrasch@posteo.de>

araujoms committed:
require one based indexing

araujoms committed:
simple optimization

araujoms committed:
still better than the fallback

dkarrasch commented on test/matmul.jl:
```diff
@@ -529,6 +529,22 @@ end
+        a = randn(T, 3, 4)
+        csmall = similar(a, 3, 3)
+        cbig = similar(a, 4, 4)
+        LinearAlgebra._generic_matmatmul!(csmall, a, a', true, false)
```
Given the non-Blas-eltypes, this (and below) could call `mul!` and dispatch should lead it to where we want it. If it doesn't, we have another issue.

araujoms replied:
I did test that it was calling the generic fallback. Are you suggesting that I should call `_generic_matmatmul_generic!` directly?

dkarrasch replied:
No, I was suggesting to call `mul!`.

araujoms replied:
I still don't understand. `mul!` will call my code, and the test becomes a tautology.

dkarrasch replied:
Sorry, I misread. I thought you were calling your internal function. My bad!

dkarrasch reviewed (approved):
Up to a small suggestion for the test, this LGTM.

dkarrasch commented on src/matmul.jl:
```diff
@@ -570,6 +570,86 @@ Base.@constprop :aggressive function _symm_hemm_generic!(C, tA, tB, A, B, alpha,
+    @inbounds if !conjugate
+        if aat
+            for k ∈ 1:n, j ∈ 1:m
+                αA_jk = α * A[j, k]
```
In the non-commutative case, we agreed to interpret the 5-arg-mul as having alpha (and beta) being multiplied from the right.

```suggestion
                αA_jk = A[j, k] * α
```

dkarrasch commented on src/matmul.jl:
```diff
@@ -570,6 +570,86 @@ Base.@constprop :aggressive function _symm_hemm_generic!(C, tA, tB, A, B, alpha,
+                for k ∈ 2:m
+                    temp += A[k, i] * A[k, j]
+                end
+                C[i, j] += α * temp
```
```suggestion
                C[i, j] += temp * α
```

dkarrasch commented on src/matmul.jl:
```diff
@@ -570,6 +570,86 @@ Base.@constprop :aggressive function _symm_hemm_generic!(C, tA, tB, A, B, alpha,
+    else
+        if aat
+            for k ∈ 1:n, j ∈ 1:m
+                αA_jk_bar = α * conj(A[j, k])
```
```suggestion
                αA_jk_bar = conj(A[j, k]) * α
```

dkarrasch commented on src/matmul.jl:
```diff
@@ -570,6 +570,86 @@ Base.@constprop :aggressive function _symm_hemm_generic!(C, tA, tB, A, B, alpha,
+                for i ∈ 1:j-1
+                    C[i, j] += A[i, k] * αA_jk_bar
+                end
+                C[j, j] += α * abs2(A[j, k])
```
```suggestion
                C[j, j] += abs2(A[j, k]) * α
```

dkarrasch commented on src/matmul.jl:
```diff
@@ -570,6 +570,86 @@ Base.@constprop :aggressive function _symm_hemm_generic!(C, tA, tB, A, B, alpha,
+                    for k ∈ 2:m
+                        temp += conj(A[k, i]) * A[k, j]
+                    end
+                    C[i, j] += α * temp
```
```suggestion
                    C[i, j] += temp * α
```

dkarrasch commented on src/matmul.jl:
```diff
@@ -570,6 +570,86 @@ Base.@constprop :aggressive function _symm_hemm_generic!(C, tA, tB, A, B, alpha,
+                for k ∈ 2:m
+                    temp += abs2(A[k, j])
+                end
+                C[j, j] += α * temp
```
```suggestion
                C[j, j] += temp * α
```

dkarrasch reviewed (commented):
Perhaps we should have a Quaternion test case?

araujoms committed:
multiply from the right

Co-authored-by: Daniel Karrasch <daniel.karrasch@posteo.de>

araujoms committed:
no benefit to calling the generic version here

araujoms committed:
add test for Quaternions

araujoms committed:
src/matmul.jl

araujoms commented:
Ok, I've added a Quaternion test. I also added a pseudo docstring since the non-commutativity is not obvious. I assume a pseudo docstring is more appropriate than a real one since this function is not going to be exported.

jishnub commented:
Btw there's no reason to not have docstrings anymore for internal functions, now that we have `public`.

araujoms committed:
add docstring

araujoms commented:
Ok, added a docstring then.

araujoms committed:
typo

araujoms committed:
fix non-real α

dkarrasch committed:
Merge branch 'master' into genericsyrk2

dkarrasch reviewed (approved):
This is nice! I only wonder if we could "hang the generic code lower" in the dispatch hierarchy, follow the existing dispatch for longer and thereby avoid some code duplication. But this could be explored in another PR once this is in.

araujoms commented:
It can be done. I'd remove the restriction `T<:BlasFloat` from `generic_matmatmul_wrapper!`, and add methods to `syrk_wrapper!`, `herk_wrapper!`, and `gemm_wrapper!` for generic types.

I'd also like to remove the line `matmul2x2or3x3_nonzeroalpha!(C, tA, tB, A, B, α, β) && return C` from `generic_matmatmul_wrapper!`, because that breaks the feature of syrk/herk that the result is always symmetric/Hermitian.

dkarrasch commented:
So what do you think, should we merge and then you pick it up as time permits, or do you feel like you wanna work on it now?

araujoms commented:
I'd rather do it now.

jishnub commented:
> I'd also like to remove the line matmul2x2or3x3_nonzeroalpha!(C, tA, tB, A, B, α, β) && return C from generic_matmatmul_wrapper!

This line dispatches to the optimized methods for 2x2 and 3x3 matrices. Removing this would lead to dispatching to BLAS calls, which are suboptimal for small matrices.

dkarrasch commented:
> It can be done. I'd remove the restriction `T<:BlasFloat` from `generic_matmatmul_wrapper!`, and add methods to `syrk_wrapper!`, `herk_wrapper!`, and `gemm_wrapper!` for generic types.

That's what I had in mind.

> This line dispatches to the optimized methods for 2x2 and 3x3 matrices. Removing this would lead to dispatching to BLAS calls, which are suboptimal for small matrices.

If we move the generic code further down the call path, this might be handled by methods up in the call path?

araujoms commented:
What about calling my generic code instead of BLAS for small matrices?

araujoms committed:
hang generic code lower in the dispatch hierarchy

araujoms committed:
typos

araujoms commented:
Done. Moved `matmul2x2or3x3_nonzeroalpha!` down for `gemm_wrapper!`, and replaced it with my generic code for `syrk_wrapper!` and `herk_wrapper!`.

jishnub commented on src/matmul.jl:
```diff
@@ -748,12 +825,26 @@ Base.@constprop :aggressive function herk_wrapper!(C::Union{StridedMatrix{T}, St
                 stride(A, 1) == stride(C, 1) == 1 &&
-                _fullstride2(A) && _fullstride2(C))
+                _fullstride2(A) && _fullstride2(C)) &&
+                max(nA, mA) ≥ 4
```
Do we still need this condition, given that we already check for small matrices earlier? The other conditions may be determined at compile time, so we should be able to remove the branch if we are not checking for the size.

araujoms replied:
I removed the earlier check for small matrices, so we need it here.

jishnub commented on src/matmul.jl:
```diff
@@ -570,6 +569,70 @@ Base.@constprop :aggressive function _symm_hemm_generic!(C, tA, tB, A, B, alpha,
+"""
+    generic_syrk!(C::StridedMatrix{T}, A::StridedVecOrMat{T}, conjugate::Bool, aat::Bool, α, β) where {T<:Number}
+
+Computes syrk/herk for generic number types. If `conjugate` is false computes syrk, i.e.,
```
The docstring should probably say that this function only computes the upper triangular half?
```julia
julia> LinearAlgebra.generic_syrk!(zeros(3,3), ones(3,3), false, true, 1.0, 1.0)
3×3 Matrix{Float64}:
 3.0  3.0  3.0
 0.0  3.0  3.0
 0.0  0.0  3.0
```

araujoms replied:
Sure, done.

araujoms committed:
update docstring

jishnub commented on src/matmul.jl:
```diff
@@ -713,12 +776,26 @@ Base.@constprop :aggressive function syrk_wrapper!(C::StridedMatrix{T}, tA::Abst
+        else
+            return copytri!(generic_syrk!(C, A, false, tA_uc == 'N', alpha, beta), 'U')
         end
     end
```
```suggestion
            BLAS.syrk!('U', tA, alpha, A, beta, C)
        else
            generic_syrk!(C, A, false, tA_uc == 'N', alpha, beta)
        end
        copytri!(C, 'U')
        return C
    end
```
I'd suggest this small change to make the branches a little simpler, and a similar change to `herk_wrapper!`

araujoms replied:
sure, done

araujoms committed:
Update src/matmul.jl

Co-authored-by: Jishnu Bhattacharya <jishnub.github@gmail.com>

araujoms committed:
update herk for consistency

jishnub commented:
Looks fine to me, so let's merge barring any final comments

araujoms commented:
Thank you and @dkarrasch for all the feedback. Any chance of backporting this into 1.12? It could be considered as fixing a performance bug.

jishnub commented:
Performance enhancements are typically not backported.

dkarrasch commented:
This has been open and under review for quite some time. When did we branch off? Could this be viewed as a late contribution to v1.12? @ViralBShah

jishnub commented:
cc @KristofferC

jishnub commented:
Given that #1297 is a bugfix to this PR, I would suggest that we don't backport this. Adding features often introduces bugs inadvertently that may then delay releases.

araujoms commented:
That's very embarrassing. I realized when writing that PR that it was a strong argument against backporting, but I couldn't leave the bug unfixed.

For what it's worth I found the bug when doing a final review of the code to make sure it was safe to backport. I'm sure everything's fine now.

I'd be glad if at least #1247 could be backported, though, that is a straightforward bugfix.

jishnub commented:
I've marked it for backport.
